File: Ejercicios/Java/androidapp-master/S08-api-minitwitter/CL21-AndroidJetpackRepository/app/src/main/java/com/androidavanzado/cl05_minitwitter/retrofit/AuthTwitterClient.java
package com.androidavanzado.cl05_minitwitter.retrofit;

import com.androidavanzado.cl05_minitwitter.common.Constantes;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class AuthTwitterClient {
    private static AuthTwitterClient instance = null;
    private AuthTwitterService miniTwitterService;
    private Retrofit retrofit;

    public AuthTwitterClient() {
        // Incluir en la cabecera de la petición el TOKEN que autoriza al usuario
        //El interceptor lo usaremos para adjuntar a cada peticion la informaicon del token
        //en la cabecera del interceptor
        OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder();
        okHttpClientBuilder.addInterceptor(new AuthInterceptor());
        OkHttpClient cliente = okHttpClientBuilder.build();

        retrofit = new Retrofit.Builder()
                .baseUrl(Constantes.API_MINITWITTER_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(cliente)
                .build();

        miniTwitterService = retrofit.create(AuthTwitterService.class);
    }

    // Patrón Singleton
    public static AuthTwitterClient getInstance() {
        if(instance == null) {
            instance = new AuthTwitterClient();
        }
        return instance;
    }

    public AuthTwitterService getAuthTwitterService() {
        return miniTwitterService;
    }


}

File: Ejercicios/Mini-Twitter/app/src/main/java/com/example/minitwitter/retrofit/api/AuthTwitterClient.kt
package com.example.minitwitter.retrofit.api

import android.content.Context
import com.example.minitwitter.utils.Constantes
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class AuthTwitterClient(val context:Context) {
    private var authTwitterService: AuthTwitterService
    private val retrofit: Retrofit
    private var instace:AuthTwitterClient? = null

    init { // Incluir en la cabecera de la petición el TOKEN que autoriza al usuario
        val okHttpClientBuilder = OkHttpClient.Builder()
        okHttpClientBuilder.addInterceptor(AuthIntercpetor(context))
        val cliente = okHttpClientBuilder.build()
        retrofit = Retrofit.Builder()
            .baseUrl(Constantes.URL_BASE)
            .addConverterFactory(GsonConverterFactory.create())
            .client(cliente)
            .build()
        authTwitterService = retrofit.create(AuthTwitterService::class.java)
    }

    fun getInstace():AuthTwitterClient{
        if(instace == null) instace = AuthTwitterClient(context)
        return instace as AuthTwitterClient
    }

    fun getAuthTwitterService():AuthTwitterService = authTwitterService

}